### ISSUE TITLE
posewarper stage1 checkpoint loading problem

### ISSUE THREAD

**Issue by zwfcrazy** [open-mmlab/mmpose#1362]
I tried to load posewarper stage1 checkpoint:
hrnet_w48_posetrack18_384x288_posewarper_stage1-08b632aa_20211130.pth
with the config file:
configs/body/2d_kpt_sview_rgb_vid/posewarper/posetrack18/hrnet_w48_posetrack18_384x288_posewarper_stage1.py

But the checkpoint does not fit the config file as:

> The model and loaded state dict do not match exactly

> unexpected key in source state_dict: neck.trans_layer.weight, neck.trans_layer.bias

> missing keys in source state_dict: keypoint_head.final_layer.weight, keypoint_head.final_layer.bias

**Comment by liqikai9** [comment]
> hrnet_w48_posetrack18_384x288_posewarper_stage1-08b632aa_20211130.pth

Hi. This checkpoint is used for **stage2** training. To train the **stage1** with the config file: hrnet_w48_posetrack18_384x288_posewarper_stage1.py, you do not have to load this checkpoint. 

You can refer to this [link ](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_vid/posewarper/posetrack18/hrnet_posetrack18_posewarper.md)for more detail.

**Comment by zwfcrazy** [comment]
> 

Hi, actually I want to compare stage1 and stage2 in order to see how much improvement stage2 can make. So I want to test stage1, not to train stage2.

**Comment by zwfcrazy** [comment]
> > hrnet_w48_posetrack18_384x288_posewarper_stage1-08b632aa_20211130.pth
> 
> Hi. This checkpoint is used for **stage2** training, which is used to initialize the weights of the model before `stage2` training.
> 
> To train the **stage1** with the config file: `hrnet_w48_posetrack18_384x288_posewarper_stage1.py`, you do not have to load this checkpoint. The checkpoint needed is already specified in the `load_from` field in the [config](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_vid/posewarper/posetrack18/hrnet_w48_posetrack18_384x288_posewarper_stage1.py#L5).
> 
> You can refer to this [link ](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_vid/posewarper/posetrack18/hrnet_posetrack18_posewarper.md)for more detail.

Hi, actually I want to compare stage1 and stage2 in order to see how much improvement stage2 can make. So I want to test stage1, not to train stage2.

**Comment by liqikai9** [comment]
I see. As stage1 and stage2 have different model architectures, the names of the layer are slightly different, which caused your issue. Specifically, the last layer in the stage1 model has the name `keypoint_head.final_layer`, while in the stage2 model, the same layer has the name `neck.trans_layer`. 

So if you want to load the stage1 checkpoint file with the stage1 config file:
`configs/body/2d_kpt_sview_rgb_vid/posewarper/posetrack18/hrnet_w48_posetrack18_384x288_posewarper_stage1.py`, 
you can load the stage1 checkpoint file we provide and **manually modify** the names of the last two keys, from `neck.trans_layer.weight, neck.trans_layer.bias` into `keypoint_head.final_layer.weight, keypoint_head.final_layer.bias` accordingly, then save it as the new checkpoint file you need.

After all, the checkpoint file has the data structure of **`OrderedDict`**, where the keys are corresponding to the names of the network layers, and the values to the corresponding weights. The checkpoint can be modified easily.

The reason we do not provide the version of the checkpoint you need is that this checkpoint is originally prepared for stage2 to initialize the weights, so the names for changed into stage2 style for convenience. 

Further, if you want to compare these two stages, as the stage1 model is trained in a single-frame setting, we already have the similar config file [here](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/posetrack18/hrnet_w48_posetrack18_384x288.py). This config is mostly the same as the stage1 config, except that they have different training epochs and learning rate settings. The results can be found [here](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/posetrack18/hrnet_posetrack18.md). 

Hope this can help!

**Comment by zwfcrazy** [comment]
> I see. As stage1 and stage2 have different model architectures, the names of the layer are slightly different, which caused your issue. Specifically, the last layer in the stage1 model has the name `keypoint_head.final_layer`, while in the stage2 model, the same layer has the name `neck.trans_layer`.
> 
> So if you want to load the stage1 checkpoint file with the stage1 config file: `configs/body/2d_kpt_sview_rgb_vid/posewarper/posetrack18/hrnet_w48_posetrack18_384x288_posewarper_stage1.py`, you can load the stage1 checkpoint file we provide and **manually modify** the names of the last two keys, from `neck.trans_layer.weight, neck.trans_layer.bias` into `keypoint_head.final_layer.weight, keypoint_head.final_layer.bias` accordingly, then save it as the new checkpoint file you need.
> 
> After all, the checkpoint file has the data structure of **`OrderedDict`**, where the keys are corresponding to the names of the network layers, and the values to the corresponding weights. The checkpoint can be modified easily.
> 
> The reason we do not provide the version of the checkpoint you need is that this checkpoint is originally prepared for stage2 to initialize the weights, so the names for changed into stage2 style for convenience.
> 
> Further, if you want to compare these two stages, as the stage1 model is trained in a single-frame setting, we already have the similar config file [here](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/posetrack18/hrnet_w48_posetrack18_384x288.py). This config is mostly the same as the stage1 config, except that they have different training epochs and learning rate settings. The results can be found [here](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/posetrack18/hrnet_posetrack18.md).
> 
> Hope this can help!

Thanks for your reply! I didn't know the two layers are equivalent.

**Comment by zwfcrazy** [comment]
FYI
Here are my test results.
Stage 2 shows no improvement in terms of pose mAP. Maybe tracking performance is improved?

Ground truth bboxes are used in this test.
In order to make a fair comparison, frame weights are set to be equally distributed in all window settings.

Window | Head | Shou | Elb | Wri | Hip | Knee | Ankl | Total
-- | -- | -- | -- | -- | -- | -- | -- | --
Stage1 Hrnet | 88.3 | 90.7 | 86.8 | 82.2 | 81.7 | 84.5 | 82 | 85.4
[-3, -2, -1, 0, 1, 2, 3] | 88.2 | 90.3 | 86 | 81.6 | 81.8 | 83.7 | 81.4 | 84.9
[-2, -1, 0, 1, 2] | 88.2 | 90.3 | 86.1 | 81.6 | 81.8 | 83.8 | 81.4 | 85
[-1, 0, 1] | 88.1 | 90.3 | 86 | 81.5 | 81.8 | 83.7 | 81.4 | 84.9
[-3, -2, -1, 0] | 88 | 90.2 | 85.8 | 81.2 | 81.6 | 83.6 | 81 | 84.7
[-2, -1, 0] | 88.1 | 90.2 | 85.9 | 81.3 | 81.7 | 83.6 | 81.2 | 84.8
[-1, 0] | 88.1 | 90.2 | 85.9 | 81.3 | 81.6 | 83.6 | 81.2 | 84.8
[-3, -2, -1, 0, 1] | 88.2 | 90.2 | 86 | 81.5 | 81.7 | 83.7 | 81.3 | 84.9
[-2, -1, 0, 1] | 88.2 | 90.3 | 86.1 | 81.5 | 81.8 | 83.8 | 81.4 | 84.9
